apollo-router/src/services/router/batching.rs

```diff
@@ -0,0 +1,421 @@
+        &self,
+        bytes: &Bytes,
+    ) -> Result<Option<Vec<graphql::Request>>, TranslateError> {
+        // REVIEW NOTE(@goto-bus-stop): Previously, batching first attempted to parse a single
+        // response, and only attempted to parse a batch if that failed. With batching as a
+        // separate layer, we can't do that anymore (as parsing is the responsibility of a
+        // downstream service), so to avoid re-parsing in the unbatched case we need an up front
+        // check for if it's likely to be a batch.
+        let first_non_ws_character = bytes.iter().find(|byte| !byte.is_ascii_whitespace());
+        if first_non_ws_character != Some(&b'[') {
+            // Not a batch request
+            return Ok(None);
+        }
+
+        if self.config.enabled && matches!(self.config.mode, BatchingMode::BatchHttpLink) {
+            let result = graphql::Request::batch_from_bytes(bytes).map_err(|e| TranslateError {
+                status: StatusCode::BAD_REQUEST,
+                extension_code: "INVALID_GRAPHQL_REQUEST".to_string(),
+                extension_details: format!("failed to deserialize the request body into JSON: {e}"),
+            })?;
+
+            if result.is_empty() {
+                return Err(TranslateError {
+                    status: StatusCode::BAD_REQUEST,
+                    extension_code: "INVALID_GRAPHQL_REQUEST".to_string(),
+                    extension_details:
+                        "failed to decode a valid GraphQL request from path: empty array "
+                            .to_string(),
+                });
+            }
+
+            if self.config.exceeds_batch_size(&result) {
+                return Err(TranslateError {
+                    status: StatusCode::UNPROCESSABLE_ENTITY,
+                    extension_code: "BATCH_LIMIT_EXCEEDED".to_string(),
+                    extension_details: format!(
+                        "Batch limits exceeded: you provided a batch with {} entries, but the configured maximum router batch size is {}",
+                        result.len(),
+                        self.config.maximum_size.unwrap_or_default()
+                    ),
+                });
+            }
+
+            Ok(Some(result))
+        } else {
+            let extension_details = if self.config.enabled
+                && !matches!(self.config.mode, BatchingMode::BatchHttpLink)
+            {
+                format!("batching not supported for mode `{}`", self.config.mode)
+            } else {
+                "batching not enabled".to_string()
+            };
+            Err(TranslateError {
+                status: StatusCode::BAD_REQUEST,
+                extension_code: "BATCHING_NOT_ENABLED".to_string(),
+                extension_details,
+            })
+        }
+    }
+
+    /// Turns a parsed batch of queries into multiple separate requests that can flow through the
+    /// rest of the pipeline.
+    fn build_batch_requests(
```

Comment:
This contains the batching parts of what used to be `RouterService::translate_requests`.